cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,264 @@
+
+  auto& group_labels = *_group_labels;
+  thrust::fill(group_labels.begin(), group_labels.end(), 0);
+  auto group_labels_ptr = group_labels.data().get();
+  auto group_ids_ptr = group_indices().data().get();
+  thrust::for_each_n(thrust::make_counting_iterator(1),
+                    group_indices().size() - 1,
+                    permutation_label_setter{group_labels_ptr, group_ids_ptr});
+  thrust::inclusive_scan(thrust::device,
+                        group_labels.begin(),
+                        group_labels.end(),
+                        group_labels.begin());
+
+  return group_labels;
+}
+
+gdf_column const& groupby::unsorted_labels() {
+  if (_unsorted_labels)
+    return *_unsorted_labels;
+
+  _unsorted_labels = std::make_unique<gdf_column>(
+    allocate_column(gdf_dtype_of<gdf_size_type>(),
+                    key_sort_order().size));
+  auto& unsorted_labels = *_unsorted_labels;
+  cudaMemset(unsorted_labels.valid, 0,
+              gdf_num_bitmask_elements(unsorted_labels.size));
+  
+  gdf_column group_labels_col{};
+  gdf_column_view(&group_labels_col, 
+                  const_cast<gdf_size_type*>(group_labels().data().get()), 
+                  nullptr,
+                  group_labels().size(), 
+                  gdf_dtype_of<gdf_size_type>());
+  cudf::table t_sorted_labels{&group_labels_col};
+  cudf::table t_unsorted_labels{&unsorted_labels};
+  cudf::detail::scatter(&t_sorted_labels,
+                        reinterpret_cast<gdf_size_type*>(key_sort_order().data),
+                        &t_unsorted_labels);
+  return unsorted_labels;
+}
+
+
+std::pair<gdf_column, rmm::device_vector<gdf_size_type> >
+groupby::sort_values(gdf_column const& val_col) {
+  auto idx_col = allocate_column(gdf_dtype_of<gdf_index_type>(),
+                                _key_table.num_rows(),
+                                false);
+
+  // We need a table constructor that can take const initializer list
+  auto unsorted_val_col = const_cast<gdf_column*> (&val_col);
+  auto unsorted_label_col = const_cast<gdf_column*> (&unsorted_labels());
+  auto unsorted_table = cudf::table{unsorted_label_col, unsorted_val_col};
+
+  gdf_context context{};
+  context.flag_groupby_include_nulls = _include_nulls;
+  gdf_order_by(unsorted_table.begin(),
+              nullptr,
+              unsorted_table.num_columns(), // always 2
+              &idx_col,
+              const_cast<gdf_context*>(&context));
+
+  cudf::table unsorted_val_col_table{unsorted_val_col};
+  auto sorted_val_col = allocate_like(val_col, _num_keys);
+  cudf::table sorted_val_col_table{&sorted_val_col};
+  cudf::gather(&unsorted_val_col_table,
+              reinterpret_cast<gdf_size_type*>(idx_col.data),
+              &sorted_val_col_table);
+  gdf_column_free(&idx_col);
+
+  // Get number of valid values in each group
+  rmm::device_vector<gdf_size_type> val_group_sizes(group_indices().size());
+  rmm::device_vector<gdf_size_type> d_bools(sorted_val_col.size);
```

Comment:
That's how I initially did it but I ran into some issues. Let me try again and I'll post the errors i got.

Reply:
Here's the errors i got:
### Case 1: 
```c++
[col_valid] __device__ (gdf_size_type i) { 
  return (col_valid) ? bit_mask::is_valid(col_valid, i) : true;
}
```
#### Error: 
```
unknown file: Failure
C++ exception with description "reduce_by_key failed on 2nd step: invalid device function" thrown in the test body.
```
### Case 2: 
```c++
[col_valid] __host__ __device__ (gdf_size_type i) { 
  return (col_valid) ? bit_mask::is_valid(col_valid, i) : true;
}
```
#### Error: 
```
Expected: 3 4.5 7 
  Actual: 0 1 2
```
This is understandable because the iterator generates bools and boolean arithmetic (`1+1=1`) causes valid group sizes to become all 1s. What's not understandable is how adding `__host__` made the runtime error disappear.
### Case 3:
```c++
[col_valid] __device__ (gdf_size_type i) -> int { 
  return (col_valid) ? bit_mask::is_valid(col_valid, i) : true;
}
```
Works.

Reply:
Then I'd have to repeat the `thrust::reduce_by_key` call in both branches no? Those two iterators would create different instantiations of the call. Thus increasing the already high compile time for seemingly little speed benefit at runtime.

Reply:
Just checked. This adds 14s to this file which already took a whole minute.